File: epi_judge_python/is_tree_balanced.py
from test_framework import generic_test
from collections import namedtuple

node = namedtuple("node", ["height", "balanced"])
def is_balanced_binary_tree(u):
    def helper(tree):
        if not tree:
            return node(0, True)
        l_height, l_bal = helper(tree.left)
        r_height, r_bal = helper(tree.right)
        if abs(l_height - r_height) > 1 or not l_bal or not r_bal:
            return node(0, False)
        return  node(max(l_height, r_height) + 1, True)
    return helper(u).balanced


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("is_tree_balanced.py",
                                       'is_tree_balanced.tsv',
                                       is_balanced_binary_tree))

File: epi_judge_java/epi/IsTreeBalanced.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
public class IsTreeBalanced {
  private static class Pair {
    final boolean balanced;
    final int height;
    public Pair(boolean balanced, int height){
      this.balanced = balanced;
      this.height = height;
    }
  }

  public static Pair recursiveCheck(BinaryTreeNode<Integer> tree){
    if (tree == null){
      return new Pair(true, 0);
    }
    Pair left = recursiveCheck(tree.left);
    Pair right = recursiveCheck(tree.right);

    return new Pair(left.balanced && right.balanced && Math.abs(left.height - right.height) <= 1,
            Math.max(left.height, right.height) + 1 );
  }

  @EpiTest(testDataFile = "is_tree_balanced.tsv")
  public static boolean isBalanced(BinaryTreeNode<Integer> tree) {

    return recursiveCheck(tree).balanced;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "IsTreeBalanced.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
